STYLEGUIDE.md

```diff
@@ -0,0 +1,1336 @@
+      }
+    }
+
+#### Be wary of null
+A method should return an `Optional` to indicate the possibility of a null value.
+
+
+#### Clean up with finally
+
+    :::java
+    FileInputStream in = null;
+    try {
+      ...
+    } catch (IOException e) {
+      ...
+    } finally {
+      Closeables.closeQuietly(in);
+    }
+    
+    //better - use autoclose
+    try (FileInputStream in = new FileInputStream(file)) {
+      ...
+    } catch (IOException e) {
+      ...
+    }
+
+Even if there are no checked exceptions, there are still cases where you should use try/finally
+to guarantee resource symmetry.
+
+    :::java
+    // Bad.
+    //   - Mutex is never unlocked.
+    mutex.lock();
+    throw new NullPointerException();
+    mutex.unlock();
+
+    // Good.
+    mutex.lock();
+    try {
+      throw new NullPointerException();
+    } finally {
+      mutex.unlock();
+    }
+
+    // Bad.
+    //   - Connection is not closed if sendMessage throws.
+    if (receivedBadMessage) {
+      conn.sendMessage("Bad request.");
+      conn.close();
+    }
+
+    // Good.
+    if (receivedBadMessage) {
+      try {
+        conn.sendMessage("Bad request.");
+      } finally {
+        conn.close();
+      }
+    }
+
+
+### Clean code
+
+#### Disambiguate
+Favor readability - if there's an ambiguous and unambiguous route, always favor unambiguous.
+
+    :::java
+    // Bad.
+    //   - Depending on the font, it may be difficult to discern 1001 from 100l.
+    long count = 100l + n;
+
+    // Good.
+    long count = 100L + n;
+
+#### Remove dead code
+Delete unused code (imports, fields, parameters, methods, classes).  They will only rot.
+
+#### Use specific abstract types to declare method return types
+When declaring method return types it is best to strive to be specific yet abstract. The reason is that you don't know what will the callers of the method need. You don't want to degrade functionality too much. The callers can always abstract the returned type farther if needed.
+
+    :::java
+    // Bad.
+    //   - Implementations of Database must match the ArrayList return type. We need to depend on abstractions not concretions.
+    //   - Changing return type to Set<User> or List<User> could break implementations and users.
+    interface Database {
+      ArrayList<User> fetchUsers(String query);
+    }
+
+    //Good
+    // Here we stress that we reurn a list of users in a certain order 
+     interface Database {
+      List<User> fetchUsers(String query);
+    }
+    
+     //Good
+    // Here we stress that we reurn a list of unique users in no particular order. If order matters use NavigableSet
+     interface Database {
+      Set<User> fetchUsers(String query);
+    }
+
+    // Good.
+    //   - Iterable defines the minimal functionality required of the return.
+    //An advantage here is that you have more possibilities to change the implementation. Notice that the program shouldn't break no matter what implementing subclass you use. 
```

Comment:
this line is too long and not easily reviewable on GH :)

Reply:
:)